File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/intake/Spin.java
package org.firstinspires.ftc.teamcode.intake;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.Servo;
import org.firstinspires.ftc.teamcode.utils.ManualMechanismTeleOp;
import org.firstinspires.ftc.teamcode.utils.ServoPositionMechanism;

import static org.firstinspires.ftc.teamcode.Positions.Intake.Spin.middle;

/**
 * Class representing the horizontal rotation of the intake claw.
 * It extends the ServoPositionMechanism class to control the rotation servo.
 */
public class Spin extends ServoPositionMechanism {
    private final HardwareMap hardwareMap;
    /**
     * @param hardwareMap the hardware map to get the servo from.
     */
    public Spin(HardwareMap hardwareMap) {
        super(middle);
        this.hardwareMap = hardwareMap;
        adjustMultiplier(0.005);
    }
    /**
     * Gets the servo associated with the Spin mechanism.
     * @return an array containing the rotation servo.
     */
    @Override
    protected Servo[] getServos() {
        return new Servo[]{hardwareMap.get(Servo.class, "Spin")};
    }
}

/**
 * TeleOp class for testing the Spin mechanism manually.
 */
@TeleOp(name = "Spin Test", group = "C Intake")
class SpinTest extends ManualMechanismTeleOp {
    public SpinTest() {
        super(Spin::new);
    }
}

/**
 * TeleOp class for testing the Spin mechanism positions.
 * @noinspection DuplicatedCode
 */
@TeleOp(name = "Spin Positions Test", group = "D Intake")
@Config
class SpinPositions extends LinearOpMode {
    public static double current = 0.0;
    @Override
    public void runOpMode() {
        Spin spin = new Spin(hardwareMap);
        waitForStart();
        while (opModeIsActive()) {
            if (gamepad1.dpad_left) current = middle;
            else current += spin.getMultiplier() * (gamepad1.right_trigger - gamepad1.left_trigger);
            spin.setTargetPosition(current);

            TelemetryPacket packet = new TelemetryPacket();
            if(!spin.run(packet)) requestOpModeStop();
            FtcDashboard.getInstance().sendTelemetryPacket(packet);
            telemetry.addData("Current Position", current);
            telemetry.update();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/intake/Spin.kt
package org.firstinspires.ftc.intake

import com.acmerobotics.dashboard.config.Config
import com.qualcomm.robotcore.eventloop.opmode.Disabled
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.Servo
import org.firstinspires.ftc.intake.Spin.Companion.left
import org.firstinspires.ftc.intake.Spin.Companion.middle
import org.firstinspires.ftc.intake.Spin.Companion.right
import org.firstinspires.ftc.utils.ManualMechanismTeleOp
import org.firstinspires.ftc.utils.Movement
import org.firstinspires.ftc.utils.PressAction
import org.firstinspires.ftc.utils.ServoPositionMechanism

@Config
class Spin(hardwareMap: HardwareMap) : ServoPositionMechanism(middle) {
    companion object {
        /**
         * The position of the intake spin when it is in the middle position.
         */
        @JvmField
        @Volatile
        var middle= 0.525
        /**
         * The position of the intake spin when it is in the left position.
         */
        @JvmField
        @Volatile
        var left = 0.265
        /**
         * The position of the intake spin when it is in the right position.
         */
        @JvmField
        @Volatile
        var right = 0.785
    }
    /**
     * Gets the servo associated with the Spin mechanism.
     */
    override val servos: Array<Servo> = arrayOf(hardwareMap.servo["Intake Spin"])
}
/**
 * TeleOp class for testing the Spin mechanism manually.
 */
@TeleOp(name = "Intake Spin Test", group = "C")
@Disabled
private class SpinTest : ManualMechanismTeleOp(::Spin)
/**
 * TeleOp class for testing the positions of the Spin mechanism.
 */
@TeleOp(name = "Intake Spin Position Test", group = "D")
@Disabled
private class SpinPositions : Movement() {
    /**
     * The Spin mechanism to be tested.
     */
    override lateinit var system: Spin
    override fun systemInit() {
        system = Spin(hardwareMap)

        actions.add(PressAction(gamepad1::dpad_down) { system.targetPosition = middle })
        actions.add(PressAction(gamepad1::dpad_left) { system.targetPosition = left })
        actions.add(PressAction(gamepad1::dpad_right) { system.targetPosition = right })
    }

    override fun systemLoop() {
        // No need for a loop
    }
}